source/detail/parsers.cpp

```diff
@@ -0,0 +1,766 @@
+        const char decimal_separator = std::use_facet<std::numpunct<char>>(loc).decimal_point();
+        bool tried_dot = false;
+        bool tried_comma = false;
+
+        // Step 1: try to parse the number as it is.
+        char *internal_end = nullptr;
+        T parsed = func(string, &internal_end);
+
+        if (parser_decimal_separator == '.')
+        {
+            tried_dot = true;
+        }
+        else if (parser_decimal_separator == ',')
+        {
+            tried_comma = true;
+        }
+
+        // Step 2: if the string was only partially parsed, try to replace the decimal separator by the one from the provided locale.
+        if (*internal_end == decimal_separator && parser_decimal_separator != decimal_separator)
+        {
+            std::string replaced(string);
+            std::replace(replaced.begin(), replaced.end(), decimal_separator, parser_decimal_separator);
+            parsed = func(replaced.c_str(), &internal_end);
+
+            if (decimal_separator == '.')
+            {
+                assert(!tried_dot);
+                tried_dot = true;
+            }
+            else if (decimal_separator == ',')
+            {
+                assert(!tried_comma);
+                tried_comma = true;
+            }
+
+            // The end has to point to the end of the original string, not the replaced one.
+            // The great C Standard Library is not const correct, so we can't be either... ugh. Note that you are NOT allowed to modify
+            // the pointer to the end (thus modifying the string) without resulting in undefined behaviour!
+            internal_end = const_cast<char *>(string + (internal_end - replaced.c_str()));
+        }
+
+        // Step 3: if the string was only partially parsed even when using the provided locale,
+        // maybe the locale has the wrong decimal separator and . was the decimal separator?
+        if (*internal_end == '.' && !tried_dot)
+        {
+            assert(parser_decimal_separator != '.');
+            std::string replaced(string);
+            std::replace(replaced.begin(), replaced.end(), '.', parser_decimal_separator);
+            parsed = func(replaced.c_str(), &internal_end);
+
+            tried_dot = true;
+
+            // The end has to point to the end of the original string, not the replaced one.
+            // The great C Standard Library is not const correct, so we can't be either... ugh. Note that you are NOT allowed to modify
+            // the pointer to the end (thus modifying the string) without resulting in undefined behaviour!
+            internal_end = const_cast<char *>(string + (internal_end - replaced.c_str()));
+        }
+
+        // Step 4: if the string was only partially parsed even when using the provided locale,
+        // maybe the locale has the wrong decimal separator and , was the decimal separator?
+        if (*internal_end == ',' && !tried_comma)
```

Comment:
Since we already discussed above that all serialized numbers should not contain any thousands separators and only `.` as the decimal separator, I can change this logic - see above.